File: python/trees/trees_5.py
class TreeNodeWithParent:
    def __init__(self, data=None, parent=None, left=None, right=None):
        self.data = data
        self.parent = parent
        self.left = left
        self.right = right

    def __repr__(self):
        if not self.left and not self.right:
            return str(self.data)
        return f'{self.data} => ({self.left}, {self.right})'


# O(log n)
# return the left most node of this subtree
def check_right(tn):
    while tn.left:
        tn = tn.left

    return tn


# O(log n)
# return the first ancestor whose value is greater than
# or equal to the value of the source node
def check_parent(tn):
    parent = tn.parent

    while parent:
        if parent.data >= tn.data:
            return parent
        parent = parent.parent

    return None


def find_next(tn):
    if not tn:
        return None

    if tn.right:
        return check_right(tn.right)

    return check_parent(tn)

File: go/pkg/trees/trees_5.go
package trees

import (
	"fmt"
	"strconv"
)

// TreeNodeWithParent represents a binary tree node with a reference to its parent node
type TreeNodeWithParent struct {
	Data   int
	Parent *TreeNodeWithParent
	Left   *TreeNodeWithParent
	Right  *TreeNodeWithParent
}

// ToString provides a string representation for binary tree node
func (n *TreeNodeWithParent) ToString() string {
	if n == nil {
		return "nil"
	}
	if n.Left == nil && n.Right == nil {
		return strconv.Itoa(n.Data)
	}
	return fmt.Sprintf("%d => (%s, %s)", n.Data, n.Left.ToString(), n.Right.ToString())
}

// checkRight runs in O(log n)
func checkRight(tn *TreeNodeWithParent) *TreeNodeWithParent {
	for tn.Left != nil {
		tn = tn.Left
	}

	return tn
}

// checkParent runs in O(log n)
func checkParent(tn *TreeNodeWithParent) *TreeNodeWithParent {
	parent := tn.Parent

	for parent != nil {
		if parent.Data >= tn.Data {
			return parent
		}
		parent = parent.Parent
	}

	return nil
}

// FindNext runs in O(log n)
func FindNext(tn *TreeNodeWithParent) *TreeNodeWithParent {
	if tn == nil {
		return nil
	}

	if tn.Right != nil {
		return checkRight(tn.Right)
	}

	return checkParent(tn)
}